In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path

path = r'C:\Users\Josh Wattay\OneDrive\Documents\Desktop\Flight Delay Analysis - Raw Data\2021'

In [3]:
# Import Datasets

df = pd.read_csv(os.path.join(path, 'flights2021.csv'), index_col = False)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_6528\3506775312.py:3: DtypeWarning: Columns (11,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, 'flights2021.csv'), index_col = False)


In [4]:
# Shape Check
df.shape

(6311871, 70)

In [5]:
# Set the display option to show all 70 columns
pd.set_option('display.max_columns', 70)

In [7]:
#Inspect Dataframe and begin Cleaning by ensuring columns are named appropriately
df.head()

Unnamed: 0  Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0           0  2021        1      1           2          6  2021-01-02   
1           1  2021        1      1           3          7  2021-01-03   
2           2  2021        1      1           4          1  2021-01-04   
3           3  2021        1      1           7          4  2021-01-07   
4           4  2021        1      1           8          5  2021-01-08   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             4628   
1                          DL                             4628   
2                          DL                             4628   
3                          DL                             4628   
4                          DL                             4628   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 9E   
1                                               NaN                 9E   
2                                               NaN                 9E   
3                                               NaN                 9E   
4                                               NaN                 9E   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20363                          9E      N337PQ   
1                     20363                          9E      N607LR   
2                     20363                          9E      N602LR   
3                     20363                          9E      N295PQ   
4                     20363                          9E      N324PQ   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             4628            11193             1119302   
1                             4628            11193             1119302   
2                             4628            11193             1119302   
3                             4628            11193             1119302   
4                             4628            11193             1119302   

   OriginCityMarketID Origin  OriginCityName OriginState  OriginStateFips  \
0               33105    CVG  Cincinnati, OH          KY               21   
1               33105    CVG  Cincinnati, OH          KY               21   
2               33105    CVG  Cincinnati, OH          KY               21   
3               33105    CVG  Cincinnati, OH          KY               21   
4               33105    CVG  Cincinnati, OH          KY               21   

  OriginStateName  DestAirportID  DestAirportSeqID  DestCityMarketID Dest  \
0        Kentucky          10721           1072102             30721  BOS   
1        Kentucky          10721           1072102             30721  BOS   
2        Kentucky          10721           1072102             30721  BOS   
3        Kentucky          10721           1072102             30721  BOS   
4        Kentucky          10721           1072102             30721  BOS   

  DestCityName DestState  DestStateFips  DestStateName  CRSDepTime  DepTime  \
0   Boston, MA        MA             25  Massachusetts         730    733.0   
1   Boston, MA        MA             25  Massachusetts         730    727.0   
2   Boston, MA        MA             25  Massachusetts         730    737.0   
3   Boston, MA        MA             25  Massachusetts        1715   1710.0   
4   Boston, MA        MA             25  Massachusetts        1715   1711.0   

   DepDelay  DepDelayMinutes  DepDel15  DepartureDelayGroups DepTimeBlk  \
0       3.0        

The 'Unnamed: 0' column will be removed as it is an additional index and redundant. 

In [8]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [9]:
df.shape

(6311871, 69)

In [10]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0  2021        1      1           2          6  2021-01-02   
1  2021        1      1           3          7  2021-01-03   
2  2021        1      1           4          1  2021-01-04   
3  2021        1      1           7          4  2021-01-07   
4  2021        1      1           8          5  2021-01-08   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             4628   
1                          DL                             4628   
2                          DL                             4628   
3                          DL                             4628   
4                          DL                             4628   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 9E   
1                                               NaN                 9E   
2                                               NaN                 9E   
3                                               NaN                 9E   
4                                               NaN                 9E   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20363                          9E      N337PQ   
1                     20363                          9E      N607LR   
2                     20363                          9E      N602LR   
3                     20363                          9E      N295PQ   
4                     20363                          9E      N324PQ   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             4628            11193             1119302   
1                             4628            11193             1119302   
2                             4628            11193             1119302   
3                             4628            11193             1119302   
4                             4628            11193             1119302   

   OriginCityMarketID Origin  OriginCityName OriginState  OriginStateFips  \
0               33105    CVG  Cincinnati, OH          KY               21   
1               33105    CVG  Cincinnati, OH          KY               21   
2               33105    CVG  Cincinnati, OH          KY               21   
3               33105    CVG  Cincinnati, OH          KY               21   
4               33105    CVG  Cincinnati, OH          KY               21   

  OriginStateName  DestAirportID  DestAirportSeqID  DestCityMarketID Dest  \
0        Kentucky          10721           1072102             30721  BOS   
1        Kentucky          10721           1072102             30721  BOS   
2        Kentucky          10721           1072102             30721  BOS   
3        Kentucky          10721           1072102             30721  BOS   
4        Kentucky          10721           1072102             30721  BOS   

  DestCityName DestState  DestStateFips  DestStateName  CRSDepTime  DepTime  \
0   Boston, MA        MA             25  Massachusetts         730    733.0   
1   Boston, MA        MA             25  Massachusetts         730    727.0   
2   Boston, MA        MA             25  Massachusetts         730    737.0   
3   Boston, MA        MA             25  Massachusetts        1715   1710.0   
4   Boston, MA        MA             25  Massachusetts        1715   1711.0   

   DepDelay  DepDelayMinutes  DepDel15  DepartureDelayGroups DepTimeBlk  \
0       3.0              3.0       0.0                   0.0  0700-0759   
1      -3.0     

Rename syntax for various columns for consistency and accuracy.

In [11]:
# Define the dictionary for renaming columns
rename_dict = {
    'DayofMonth': 'Day_of_Month',
    'DayOfWeek': 'Day_of_Week',
    'FlightDate': 'Flight_Date',
    'Operating_Airline': 'Operating Airline',
    'OriginAirportID': 'Origin_Airport_ID',
    'OriginAirportSeqID': 'Origin_Airport_Seq_ID',
    'OriginCityMarketID': 'Origin_City_Market_ID',
    'OriginCityName': 'Origin_City_Name',
    'OriginState': 'Origin_State',
    'OriginStateFips': 'Origin_State_Fips',
    'OriginStateName': 'Origin_State_Name',
    'DestAirportID': 'Dest_Airport_ID',
    'DestAirportSeqID': 'Dest_Airport_Seq_ID',
    'DestCityMarketID': 'Dest_City_Market_ID',
    'DestCityName': 'Dest_City_Name',
    'DestState': 'Dest_State',
    'DestStateFips': 'Dest_State_Fips',
    'DestStateName': 'Dest_State_Name',
    'CRSDepTime': 'CRS_Dep_Time',
    'DepTime': 'Dep_Time',
    'DepDelay': 'Dep_Delay',
    'DepDelayMinutes': 'Dep_Delay_Minutes',
    'DepDel15': 'Dep_Del_15',
    'DepartureDelayGroups': 'Departure_Delay_Groups',
    'DepTimeBlk': 'Dep_Time_Blk',
    'TaxiOut': 'Taxi_Out',
    'WheelsOff': 'Wheels_Off',
    'WheelsOn': 'Wheels_On',
    'TaxiIn': 'Taxi_In',
    'CRSArrTime': 'CRS_Arr_Time',
    'ArrTime': 'Arr_Time',
    'ArrDelay': 'Arr_Delay',
    'ArrDelayMinutes': 'Arr_Delay_Minutes',
    'ArrDel15': 'Arr_Del_15',
    'ArrivalDelayGroups': 'Arrival_Delay_Groups',
    'ArrTimeBlk': 'Arr_Time_Blk',
    'CancellationCode': 'Cancellation_Code',
    'CRSElapsedTime': 'CRS_Elapsed_Time',
    'ActualElapsedTime': 'Actual_Elapsed_Time',
    'AirTime': 'Air_Time',
    'DistanceGroup': 'Distance_Group',
    'CarrierDelay': 'Carrier_Delay',
    'WeatherDelay': 'Weather_Delay',
    'NASDelay': 'NAS_Delay',
    'SecurityDelay': 'Security_Delay',
    'LateAircraftDelay': 'Late_Aircraft_Delay',
    'FirstDepTime': 'First_Dep_Time',
    'TotalAddGTime': 'Total_Add_G_Time',
    'LongestAddGTime': 'Longest_Add_G_Time',
    'DivDistance': 'Div_Distance'
}

# Rename the columns
df = df.rename(columns=rename_dict)

In [12]:
df.shape

(6311871, 69)

In [13]:
# Drop the additional irrelevant columns - please refer to the 'Cleaning 2019 Flight Data' script for more details.

df = df.drop(columns=['First_Dep_Time', 'Total_Add_G_Time', 'Longest_Add_G_Time', 'Div_Distance'])

Fortunately, this raw data comes with a Duplicate flag column.
I will check the counts for this column to check for any values of 'Y' or Yes.
To be certain, I will also run a duplicate check with the .duplicated command.

In [14]:
df['Duplicate'].value_counts()

Duplicate
N    6311871
Name: count, dtype: int64

In [15]:
# So far so good, now to check for duplicates in the entire dataframe by creating a subset of duplicated rows

# Create a DataFrame containing only the duplicate rows
dups = df[df.duplicated(keep=False)]

# Display the 'dups' DataFrame
print(dups)

Empty DataFrame
Columns: [Year, Quarter, Month, Day_of_Month, Day_of_Week, Flight_Date, Marketing_Airline_Network, DOT_ID_Marketing_Airline, IATA_Code_Marketing_Airline, Flight_Number_Marketing_Airline, IATA_Code_Originally_Scheduled_Code_Share_Airline, Operating_Airline , DOT_ID_Operating_Airline, IATA_Code_Operating_Airline, Tail_Number, Flight_Number_Operating_Airline, Origin_Airport_ID, Origin_Airport_Seq_ID, Origin_City_Market_ID, Origin, Origin_City_Name, Origin_State, Origin_State_Fips, Origin_State_Name, Dest_Airport_ID, Dest_Airport_Seq_ID, Dest_City_Market_ID, Dest, Dest_City_Name, Dest_State, Dest_State_Fips, Dest_State_Name, CRS_Dep_Time, Dep_Time, Dep_Delay, Dep_Delay_Minutes, Dep_Del_15, Departure_Delay_Groups, Dep_Time_Blk, Taxi_Out, Wheels_Off, Wheels_On, Taxi_In, CRS_Arr_Time, Arr_Time, Arr_Delay, Arr_Delay_Minutes, Arr_Del_15, Arrival_Delay_Groups, Arr_Time_Blk, Cancelled, Cancellation_Code, Diverted, CRS_Elapsed_Time, Actual_Elapsed_Time, Air_Time, Flights, Distance,

In [16]:
# No dups!

Now I will address missing and/or null values.

In [17]:
# Set the display option to show all rows
pd.set_option('display.max_rows', None)

In [18]:
# Find the count of null values in each column
null_counts = df.isnull().sum()

In [19]:
# Display the count of null values
print(null_counts)

Year                                                       0
Quarter                                                    0
Month                                                      0
Day_of_Month                                               0
Day_of_Week                                                0
Flight_Date                                                0
Marketing_Airline_Network                                  0
DOT_ID_Marketing_Airline                                   0
IATA_Code_Marketing_Airline                                0
Flight_Number_Marketing_Airline                            0
IATA_Code_Originally_Scheduled_Code_Share_Airline    6311384
Operating_Airline                                          0
DOT_ID_Operating_Airline                                   0
IATA_Code_Operating_Airline                                0
Tail_Number                                            23303
Flight_Number_Operating_Airline                            0
Origin_Airport_ID       

In [20]:
# The IATA_Code_Originally_Scheduled_Code_Share column is nearly completely null
# Given that the Bureau of Transportation Statistics (BTS.gov) stated that this column shouldn't be 
# used for analysis, and it is completely null for 2019, I will be dropping it and keeping the Unique_Code
# column as suggested by BTS.gov
# Here is the language from BTS.gov:
# ""Code assigned by IATA and commonly used to identify a carrier.""
#""As the same code may have been assigned to different carriers over time, the code is not always unique.""
#""For analysis, use the Unique Carrier Code.""

# Drop the 'IATA_Code_Originally_Scheduled_Code_Share_Airline' column
df = df.drop(columns=['IATA_Code_Originally_Scheduled_Code_Share_Airline'])

In [21]:
df.shape

(6311871, 64)

There are a significant amount of null values for Dep_Delay, Taxi_Out, Taxi_In, Arr_Time.
I believe these are null values because the flights were cancelled.
I will check value counts for the Cancelled column. Value counts for '1' should be close if not the same to the values missing.

In [22]:
df['Cancelled'].value_counts()

Cancelled
0.0    6200853
1.0     111018
Name: count, dtype: int64

111,018 flights were cancelled in 2021.

This is fairly close to the missing values in:

Dep_Time                                              108325
Dep_Delay                                             108413
Dep_Delay_Minutes                                     108413
Dep_Del_15                                            108413
Departure_Delay_Groups                                10
Taxi_Out                                              110353
Wheels_Off                                            110353
Wheels_On                                             112425
Taxi_In                                               112
Arr_Time                                              112408
Arr_Delay                                             126001
Arr_Delay_Minutes                                     126001
Arr_Del_15                                            126001
Arrival_Delay_Groups                                  1260014258413

The missing values in:

Cancellation_Code                         6200853    4721342 -- are accounted for because it is the same as the tota6311871ts (5022397) minus the cancelle111,018ts (301,055) 6200853 is 4721342.6200853ore 4721342 flights have no cancellation code because they were not cancelled.

After careful consideration, I believe it is best to create a cancelled flights flag in order to prevent confusion in future analysis for delays v. cancellation. 

In [23]:
# Identify rows where flights were cancelled

cancelled_flights = df['Cancelled'] == 1

In [24]:
# Create a column to indicate if the data is missing due to cancellation

df['Missing_Due_To_Cancellation'] = cancelled_flights

Now I will address columns containing mixed datatypes.

In [25]:
df.dtypes

Year                                 int64
Quarter                              int64
Month                                int64
Day_of_Month                         int64
Day_of_Week                          int64
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int64
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int64
Operating_Airline                   object
DOT_ID_Operating_Airline             int64
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int64
Origin_Airport_ID                    int64
Origin_Airport_Seq_ID                int64
Origin_City_Market_ID                int64
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                    int64
Origin_State_Name                   object
Dest_Airpor

In [26]:
# I will change many of these datatypes to reduce the amount of memory needed by reducing int64 to int8, 16, or 32 where possible.
# Additionally, I will change the Flight_Date column to a date datatype, instead of an object.
# Lets change Flight_Date first

# Ensure 'Flight_Date' column is in datetime format
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'], errors='coerce')

In [27]:
# Extract only the date part (since we don't have time in the same column)
df['Flight_Date'] = df['Flight_Date'].dt.date

In [28]:
# Verify the result
print(df.head())
print(df.dtypes)

   Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2021        1      1             2            6  2021-01-02   
1  2021        1      1             3            7  2021-01-03   
2  2021        1      1             4            1  2021-01-04   
3  2021        1      1             7            4  2021-01-07   
4  2021        1      1             8            5  2021-01-08   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             4628   
1                          DL                             4628   
2                          DL                             4628

In [29]:
# Change datatypes to maximize available memory

df = df.astype({
    'Year': 'int16',
    'Quarter': 'int16',
    'Month': 'int16',
    'Day_of_Month': 'int16',
    'Day_of_Week': 'int16',
    'DOT_ID_Marketing_Airline': 'int16',
    'Flight_Number_Marketing_Airline': 'int16',
    'DOT_ID_Operating_Airline': 'int16',
    'Flight_Number_Operating_Airline': 'int16',
    'Origin_Airport_ID': 'int16',
    'Origin_Airport_Seq_ID': 'int16',
    'Origin_City_Market_ID': 'int16',
    'Origin_State_Fips': 'int8',
    'Dest_Airport_ID': 'int16',
    'Dest_Airport_Seq_ID': 'int16',
    'Dest_City_Market_ID': 'int16',
    'Dest_State_Fips': 'int16',
    'CRS_Dep_Time': 'int16'
})

In [30]:
df.dtypes

Year                                 int16
Quarter                              int16
Month                                int16
Day_of_Month                         int16
Day_of_Week                          int16
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int16
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int16
Operating_Airline                   object
DOT_ID_Operating_Airline             int16
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int16
Origin_Airport_ID                    int16
Origin_Airport_Seq_ID                int16
Origin_City_Market_ID                int16
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                     int8
Origin_State_Name                   object
Dest_Airpor

In [31]:
df.shape

(6311871, 65)

In [32]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2021        1      1             2            6  2021-01-02   
1  2021        1      1             3            7  2021-01-03   
2  2021        1      1             4            1  2021-01-04   
3  2021        1      1             7            4  2021-01-07   
4  2021        1      1             8            5  2021-01-08   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             4628   
1                          DL                             4628   
2                          DL                             4628   
3                          DL                             4628   
4                          DL                             4628   

  Operating_Airline   DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                 9E                     20363                          9E   
1                 9E                     20363                          9E   
2                 9E                     20363                          9E   
3                 9E                     20363                          9E   
4                 9E                     20363                          9E   

  Tail_Number  Flight_Number_Operating_Airline  Origin_Airport_ID  \
0      N337PQ                             4628              11193   
1      N607LR                             4628              11193   
2      N602LR                             4628              11193   
3      N295PQ                             4628              11193   
4      N324PQ                             4628              11193   

   Origin_Airport_Seq_ID  Origin_City_Market_ID Origin Origin_City_Name  \
0                   5190                 -32431    CVG   Cincinnati, OH   
1                   5190                 -32431    CVG   Cincinnati, OH   
2                   5190                 -32431    CVG   Cincinnati, OH   
3                   5190                 -32431    CVG   Cincinnati, OH   
4                   5190                 -32431    CVG   Cincinnati, OH   

  Origin_State  Origin_State_Fips Origin_State_Name  Dest_Airport_ID  \
0           KY                 21          Kentucky            10721   
1           KY                 21          Kentucky            10721   
2           KY                 21          Kentucky            10721   
3           KY                 21          Kentucky            10721   
4           KY                 21          Kentucky            10721   

   Dest_Airport_Seq_ID  Dest_City_Market_ID Dest Dest_City_Name Dest_State  \
0                23526                30721  BOS     Boston, MA         MA   
1                23526                30721  BOS     Boston, MA         MA   
2                23526                30721  BOS     Boston, MA         MA   
3                23526                30721  BOS     Boston, MA         MA   
4                23526                30721  BOS     Boston, MA         MA   

   Dest_State_Fips Dest_State_Name  CRS_Dep_Time  Dep_Time  Dep_Delay  \
0               25   Massachusetts           730     733.0        3.0   
1               25   Massachusetts           730     727.0       -3.0   
2               25   Massachusetts           730     737.0        7.0   
3               25   Massachusetts          1715    1710.0       -5.0   
4               25   Massachusetts          1715    1711.0       -4.0   

   Dep_Delay_Minutes  Dep_Del_15  Departure_Delay_Groups Dep_Time_Blk  \
0                3.0         0.0                     0.0    0700-0759   
1                0.0         0.0            

In [33]:
# Save Clean 2021 Data

df.to_csv(os.path.join(path, 'clean_flights2021.csv'))